crates/runtime-table-partition/src/provider.rs

```diff
@@ -118,6 +136,136 @@ impl TableProvider for PartitionTableProvider {
         _filters: &[Expr],
         _limit: Option<usize>,
     ) -> Result<Arc<dyn ExecutionPlan>, DataFusionError> {
-        todo!()
+        Err(DataFusionError::Execution(
+            "PartitionedTableProvider::scan not implemented yet".to_string(),
+        ))
+    }
+
+    async fn insert_into(
+        &self,
+        state: &dyn Session,
+        input: Arc<dyn ExecutionPlan>,
+        insert_op: InsertOp,
+    ) -> Result<Arc<dyn ExecutionPlan>, DataFusionError> {
+        let expr = self.partition_by.first().ok_or_else(|| {
+            DataFusionError::Execution("Failed to get first partition expression".to_string())
+        })?;
+        let df_schema = DFSchema::try_from(Arc::clone(&self.schema))?;
+
+        let task_ctx = Arc::new(TaskContext::from(state));
+        let mut stream = execute_stream(Arc::clone(&input), task_ctx)?;
+        let mut execution_plans = Vec::new();
+
+        while let Some(batch) = stream.next().await {
+            let batch = batch?;
+            if batch.num_rows() == 0 {
+                continue;
+            }
+
+            let partition_groups = group_by_partition(expr, &[batch], &df_schema)?;
+
+            for (scalar_value, batch) in partition_groups {
+                let partition_key = scalar_value.to_string();
+                let table_provider = {
+                    let partitions = self.partitions.read().await;
+                    if let Some(existing_provider) = partitions.get(&partition_key) {
+                        Arc::clone(existing_provider)
+                    } else {
+                        drop(partitions);
+                        let new_provider = self
+                            .creator
+                            .create_partition(scalar_value.clone())
+                            .await
+                            .map_err(|e| DataFusionError::Execution(e.to_string()))?;
+                        let mut partitions = self.partitions.write().await;
+                        partitions.insert(partition_key.clone(), Arc::clone(&new_provider));
+                        new_provider
+                    }
+                };
+
+                let mem_exec = DataSourceExec::new(Arc::new(MemorySourceConfig::try_new(
+                    &[vec![batch]],
+                    Arc::clone(&self.schema),
+                    None,
+                )?));
+
+                let plan = table_provider
+                    .insert_into(state, Arc::new(mem_exec), insert_op)
+                    .await?;
+                execution_plans.push(plan);
```

Comment:
@kczimm - do we have any special final commit/rollback logic for partitioned insert - let's say one table provider failed to insert but other completed, how we control this?

Reply:
While I don't think we can ever completely eliminate that risk in the current design, we can minimize its impact by only calling `insert_into` once on each of the partitions and keeping that stream alive until the end, and then closing them all at once. The current implementation will commit each batch, which is dangerous.